src/libsyntax/ext/deriving/generic/mod.rs

```diff
@@ -330,6 +330,39 @@ pub fn combine_substructure<'a>(f: CombineSubstructureFunc<'a>)
     RefCell::new(f)
 }
 
+fn is_generic(ty: &ast::Ty, ty_param_names: &[ast::Name]) -> bool {
+    use visit;
+
+    struct Visitor<'a> {
+        found: bool,
+        ty_param_names: &'a [ast::Name],
+    }
+
+    impl<'a> visit::Visitor<'a> for Visitor<'a> {
+        fn visit_path(&mut self, path: &'a ast::Path, _id: ast::NodeId) {
+            match path.segments.first() {
```

Comment:
Drat, you're right, this currently compiles:

```rust
mod T {
    pub type Type = i32;
}

trait Trait {
    type Type;
}

struct Baz<T: Trait> {
    a: T::Type,
}
```

I filed #21301 to get that fixed. It'll probably not be that much of an issue in practice though, as you'd have to be crazy to write that :)